### ISSUE TITLE
Name lookup broken when calling from a lambda in a class template

### ISSUE THREAD

**Issue by asklar** [microsoft/cppwinrt#1084]
See [C3779 when using type in a class method in a template class, instead of a lambda](https://developercommunity.visualstudio.com/t/C3779-when-using-type-in-a-class-method/1617634).

## Summary
inside a struct template, have a method that passes a lambda (like a cppwinrt event handler), and in that handler use a method like `IVector<T>::Size()`, or `AutoSuggestBoxTextChangedEventArgs::Reason()`. The result is you get an error like:

```
error C3779: 'winrt::impl::consume_Microsoft_UI_Xaml_Controls_IAutoSuggestBoxTextChangedEventArgs<winrt::Microsoft::UI::Xaml::Controls::IAutoSuggestBoxTextChangedEventArgs>::Reason': a function that returns 'auto' cannot be used before it is defined
```

However, disabling conformant name lookup (by passing `/Zc:twoPhase-`) makes the build succeed; however this is not a viable workaround as it makes the code non-conformant.

Long story short: it seems like some parts of cppwinrt projections can't be used from template classes?

**Comment by sylveon** [comment]
I cannot repro the issue on VS 17.1.0 preview 1.1

**Comment by asklar** [comment]
I believe sylveon had tried this on a console app -- I can't repro in a console app either but I can in a uwp app; this is on VS 17.0.4 (retail). Maybe console apps are automatically opted in to the twoPhase- flag?

**Comment by asklar** [comment]
Sure. Here's the app I posted to the devcom ticket: https://aka.ms/dc/file?name=Bb4e780d18c0f45a8a557e2487c518d5d637755777511404865_App5.zip&tid=b4e780d18c0f45a8a557e2487c518d5d637755777511404865

**Comment by kennykerr** [comment]
Not very minimal... 😉

**Comment by asklar** [comment]
it's the "file -> new uwp project" template, with like 5 lines added to mainpage :) so pretty minimal :)

**Comment by kennykerr** [comment]
It sounds like a compiler issue so I'll close this issue since you have opened a bug with the C++ team. Note that I cannot reproduce the problem since the project fails to build. I would also suggest trying to find a more minimal repro, since compiler devs are even less likely to be able reproduce the problem given the mountain of dependencies that such a project implies. Here are some tips for creating a good repro:

https://docs.microsoft.com/en-us/cpp/overview/how-to-report-a-problem-with-the-visual-cpp-toolset?view=msvc-170
